sdks/go/pkg/beam/runners/universal/extworker/extworker.go

```diff
@@ -106,12 +106,16 @@ func (s *Loopback) StopWorker(ctx context.Context, req *fnpb.StopWorkerRequest)
 // Stop terminates the service and stops all workers.
 func (s *Loopback) Stop(ctx context.Context) error {
 	s.mu.Lock()
-	defer s.mu.Unlock()
 
 	log.Infof(ctx, "stopping Loopback, and %d workers", len(s.workers))
 	s.workers = map[string]context.CancelFunc{}
 	s.lis.Close()
 	s.rootCancel()
+
+	// There can be a deadlock between the StopWorker RPC and GracefulStop
+	// which waits for all RPCs to finish, so it must be outside the critical section.
+	s.mu.Unlock()
```

Comment:
Does this introduce a new potential race where:

1) We take the lock
2) We close the LIstener
3) We release the lock
4) A worker tries to send information to the port the Listener was set up on

That seems less severe than the current race, but I'm not really sure how we'd behave there - would that cause errors? Is there a reason we need to close the listener before cancelling/stopping?

Reply:
It does not. And looking at the Stop and GracefulStop implementations, those already handle closing the listener, so I've removed that unnecessary line.

There was a race during the Stop call if a request for a new Worker started up came in, would have happily started one, so now I have it return an error in that case.

I've run the new runner's suite of pipelines 1000 times with both the race detector on and off (so that over 10k runs of this sequence), and these changes avoid the occasional flake I was seeing.  This along with earlier good goroutine hygene made them easier to find in the timeout dumps.